pkg/k8s: Synchronize heartbeat goroutine

aditighag opened the pull request:
When the heartbeat timeout is set to a low value (e.g., in tests), there's a risk that the goroutine responsible for sending heartbeat probes may not start before the timeout expires. To address this, introduce a channel to synchronize goroutine startup.

A WaitGroup could also be used, but it risks blocking the controller indefinitely if the goroutine fails to start or blocks on I/O.

Fixes: https://github.com/cilium/cilium/issues/39428

marseel commented on pkg/k8s/client/cell.go:
```diff
@@ -367,10 +367,15 @@ func runHeartbeat(logger *slog.Logger, heartBeat func(context.Context) error, ti
-	defer cancel()
 	go func() {
+		ctx, cancel = context.WithTimeout(context.Background(), timeout)
+		start <- struct{}{}
```
As far as I can see, `K8sHeartbeatTimeout` is set to `5ms` in that test. 
This also controls how often heartbeat is executed here: https://github.com/cilium/cilium/blob/afcb0c986df5b45d0037cc768c7da4f5ecc367ab/pkg/k8s/client/cell.go#L306
Client-go rate-limits number of requests to external server, be default we have 10 qps / 20 burst. 
I've checked that with:
```
	t.Log(clientset.Config().K8sClientQPS)
	t.Log(clientset.Config().K8sClientBurst)
```
in the test.
That means, if test runs for ~200-300 ms, we saturate limit from heartbeat and all consecutive calls are throttled, resulting in timeouts in heartbeat. 
This can be easily simulated by adding `time.Sleep(10 * time.Second)` after `hive.Start` here: https://github.com/cilium/cilium/blob/2b754de5808a29fee75d93a8ac7d5a5e97c34141/pkg/k8s/client/client_test.go#L323 and the test fails reliably.

I would propose to change `K8sHeartbeatTimeout` in the test from `5ms` to `150ms` instead (so 10 QPS can be sustained), or just increase client-go rate limits :)

aditighag replied:
>This can be easily simulated by adding time.Sleep(10 * time.Second) after hive.Start here:

Well spotted! 
~I noticed that this particular test has a bunch of different values set for heart beat timeout, and as a result corresponding asserts. 
I'm hoping we can transition to synctest once it's available as a follow-up.~

Edit: <errnocoffee> Nvm, I mixed this up with the other test case. 

For now, I'll increase the QPS as the test case is supposed to exercise mainly time aspect of the heartbeat testing logic.

aditighag replied:
Done! Added an artificial delay of 10s, and wasn't able to reproduce the issue caused due to throttling.

marseel replied:
Two more nits:
I see you've added `flags.Set(option.K8sHeartbeatTimeout, "150ms")` in one test, which contradicts
`For now, I'll increase the QPS as the test case is supposed to exercise mainly time aspect of the heartbeat testing logic.`. Just want to make sure this is what you wanted :) I'm fine with that change either way. 

I do not think that your previous change helps though, I would remove it - groutine was most likely starting, it was just waiting on response. `start <- struct{}{}` can preempt go-routine anyway as it's blocking operation and channel is non-buffered - depending on execution order of inner and outer goroutines. It doesn't break anything either, but it just makes code a bit harder to read/understand.

aditighag replied:
Oh! I had edited the previous note after pushing the last revision of changes, but somehow I don't see the edits. 😕 
Anyway, yes, I decided to bump up the heart beat timeout as well to prevent any further flakes. 
I had also commented saying that I would keep the first commit as the main thread *should* block on the goroutine sending heart beat probes. Also, notice that earlier the context timeout was started outside of the goroutine. I don't have a strong preference about the channel based synchronization, but I think we should at least start the timeout when the goroutine runs. We could also make the channel buffered with capacity 1 to avoid getting the goroutine preempted by the schedular. 

Edit: I thought more about this -- we can merge the 2nd commit bumping up QPS and heartbeat timeout and see how that plays out, and drop the 1st commit. It's less likely that go routine scheduler delay would cause such failures in production environments.

marseel replied:
> Edit: I thought more about this -- we can merge the 2nd commit bumping up QPS and heartbeat timeout and see how that plays out, and drop the 1st commit. It's less likely that go routine scheduler delay would cause such failures in production environments.

yea, fully agree. if scheduler takes more than couple of ms then there are probably other problems already, but still it wouldn't affect production due to higher timeout. 
I guess you might want to increase heartbeat timeout for `Test_client` test as well?

aditighag committed:
pkg/k8s: Set higher QPS limit and timeout in client_test

The test cases exercise timing aspects so
set a higher limit to avoid spurious failures. This
fixes the flakes seen in CI.

Suggested-by: Marcel Zieba <marcel.zieba@isovalent.com>
Signed-off-by: Aditi Ghag <aditi@cilium.io>
